src/System.Private.CoreLib/shared/System/AppContext.cs

```diff
@@ -17,14 +17,14 @@ public static partial class AppContext
         private static Dictionary<string, bool>? s_switches;
         private static string? s_defaultBaseDirectory;
 
-        public static string? BaseDirectory
+        public static string BaseDirectory
         {
             get
             {
                 // The value of APP_CONTEXT_BASE_DIRECTORY key has to be a string and it is not allowed to be any other type. 
                 // Otherwise the caller will get invalid cast exception
                 return (string?)GetData("APP_CONTEXT_BASE_DIRECTORY") ??
-                    (s_defaultBaseDirectory ?? (s_defaultBaseDirectory = GetBaseDirectoryCore()));
+                    (s_defaultBaseDirectory ?? (s_defaultBaseDirectory = GetBaseDirectoryCore())) ?? string.Empty;
```

Comment:
I assume this is something you and @jkotas discussed and agreed on?

Reply:
(We have not discussed this one...)

It is a corner case that is impossible to hit in .NET Core today. It will be hit in future, e.g. for WebAssembly where the file system does not exist and the path returning APIs cannot return anything useful. The existing conventions for APIs that return paths are mixed bag. `Assembly.Location` has a precedent with returning empty string, so it should be ok to return an empty string here as well.

Reply:
I wasn't sure if we wanted to just return empty string every time we got a null or if it made sense to change `GetBaseDirectory` to return empty string and stop re-evaluating it every time, is that alternative better?

Reply:
> change GetBaseDirectory to return empty string and stop re-evaluating it every time

Yes, that would be better.

Reply:
Fixed.